course3/UniquePtr/ptr_test.cc

```diff
@@ -0,0 +1,73 @@
+#include "my_unique_ptr.h"
+#include <array>
+#include <string>
+#include <cassert>
+#include <iostream>
+#include <cstring>
+#include <memory>
+
+
+struct Foo
+{
+    Foo()      { std::cout << "Foo::Foo\n";  }
+    ~Foo()     { std::cout << "Foo::~Foo\n"; }
+    void bar() { std::cout << "Foo::bar\n";  }
+    int ret6() { std::cout << "Foo::ret6\n"; return 6; }
+    std::array<int, 5> retArray(){
+        std::array<int, 5> arr;
+        for(int i=0; i<5; ++i) {
+            arr[i]=i*2;
+        }
+    return arr;
+    }
+};
+
+
+std::string s(const Foo &){
+    std::cout << "s(const Foo&)\n";
+    return "aString";
+}
+
+
+int main()
+{
+    
+    std::cout << "Testing std::unique_ptr..." << std::endl;
+    std::unique_ptr<Foo> std_p1(new Foo);  // p1 owns Foo
+    std::unique_ptr<Foo> std_p2(std::move(std_p1));  // now p2 owns Foo
+    assert(std_p1->ret6() == std_p2->ret6());
+    assert( s(*std_p1) == "aString" );
+    std_p1 = std::move(std_p2);  // ownership returns to p1
+    std::cout << "destroying std_p2...\n";
+    std::unique_ptr<Foo> std_p3;
+    assert (std_p3 == nullptr);
+    std_p3.swap(std_p1);
+    assert (std_p1 != std_p3);
+    assert (std_p3 != nullptr);
+    assert (std_p1 == nullptr);
+    std::array<int, 5>  std_arr = std_p1->retArray();
+    std::cout << "std_arr[1]=" << std_arr[1] << std::endl;
+    assert(std_arr[1] == 2);
+    std::cout << "std::unique_ptr test passes" << std::endl;
+    
+    std::cout << " " << std::endl;
+    
+    std::cout << "Testing MyUniquePtr..." << std::endl;
+    MyUniquePtr<Foo> my_p1(new Foo);  // p1 owns Foo
+    MyUniquePtr<Foo> my_p2(std::move(my_p1));  // now p2 owns Foo
```

Comment:
Due to the lack of move constructor on MyUniquePtr, and the fact that (implicitly defined)  copy-constructor is still available, this simply initializes my_p2 as a copy to my_p1. Something that should never happen in a correctly implemented _unique_ptr.
As the object is already deleted, and neither of the copies will attempt to perform any management on it, there will be no double delete.